catroid/src/main/java/org/catrobat/catroid/io/BackpackSerializer.java

```diff
@@ -0,0 +1,115 @@
+/*
+ * Catroid: An on-device visual programming system for Android devices
+ * Copyright (C) 2010-2018 The Catrobat Team
+ * (<http://developer.catrobat.org/credits>)
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * An additional term exception under section 7 of the GNU Affero
+ * General Public License, version 3, is available at
+ * http://developer.catrobat.org/license_additional_term
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package org.catrobat.catroid.io;
+
+import android.util.Log;
+
+import com.google.gson.Gson;
+import com.google.gson.GsonBuilder;
+import com.google.gson.JsonIOException;
+import com.google.gson.JsonSyntaxException;
+
+import org.catrobat.catroid.common.Backpack;
+import org.catrobat.catroid.content.Script;
+import org.catrobat.catroid.content.bricks.Brick;
+
+import java.io.BufferedReader;
+import java.io.FileNotFoundException;
+import java.io.FileReader;
+import java.io.FileWriter;
+import java.io.IOException;
+
+import static org.catrobat.catroid.common.Constants.BACKPACK_DIRECTORY;
+import static org.catrobat.catroid.common.Constants.BACKPACK_FILE;
+import static org.catrobat.catroid.common.Constants.BACKPACK_IMAGE_DIRECTORY;
+import static org.catrobat.catroid.common.Constants.BACKPACK_SCENE_DIRECTORY;
+import static org.catrobat.catroid.common.Constants.BACKPACK_SOUND_DIRECTORY;
+import static org.catrobat.catroid.common.Constants.DEFAULT_ROOT_DIRECTORY;
+
+public final class BackpackSerializer {
+
+	private static final String TAG = BackpackSerializer.class.getSimpleName();
+	private static final BackpackSerializer INSTANCE = new BackpackSerializer();
+
+	private Gson backpackGson;
+
+	public static BackpackSerializer getInstance() {
+		return INSTANCE;
+	}
+
+	private BackpackSerializer() {
+		GsonBuilder gsonBuilder = new GsonBuilder().enableComplexMapKeySerialization().setPrettyPrinting();
+		gsonBuilder.registerTypeAdapter(Script.class, new BackpackScriptSerializerAndDeserializer());
+		gsonBuilder.registerTypeAdapter(Brick.class, new BackpackBrickSerializerAndDeserializer());
+		backpackGson = gsonBuilder.create();
+
+		DEFAULT_ROOT_DIRECTORY.mkdir();
+		BACKPACK_DIRECTORY.mkdir();
+		BACKPACK_SCENE_DIRECTORY.mkdir();
+		BACKPACK_IMAGE_DIRECTORY.mkdir();
+		BACKPACK_SOUND_DIRECTORY.mkdir();
```

Comment:
or not… because some of the directories may already exist.

Reply:
Well, maybe. Or maybe not. It depends on how and where it's going to be initialised -> StorageOperations.createDir(..) throws an IOException and someone would have to catch it then -> so it would be better to use it instead, but it does not really make a difference in this case, because the result is ignored anyway and there is no real error handling if writing to the storage goes wrong. Anyway it shows that someone should think about when these initialisations take place, but tbh I don't want to do that right now -> I just want to split the StorageHandler's different functionalities into different classes.